### PR TITLE
Block data callback

### PR DESCRIPTION
**Pull request by braingram**
This adds a `data_callback` argument to `asdf.block.Block` that allow the block to dynamically load data during write. This is work towards a deferred block implementation where this callback will be used during the deferred write.

### CODE REVIEW

**Comment by braingram** [comment]
Failures are unrelated and addressed in #1470

**Review by WilliamJamieson — CHANGES_REQUESTED** [review]
This needs a changelog as `Block` is a public class.

**Comment by braingram** [comment]
> This needs a changelog as `Block` is a public class.

Based on this comment I took `Block` to not be a public class:
https://github.com/asdf-format/asdf/blob/b8e1759aa9f013283942140c94732a170eb9a703/asdf/block.py#L799-L801

I do not think this should be public and would rather not include a changelog entry which might cause users to think it's public.

Are there changes that should be made to make this more explicit?

**Comment by WilliamJamieson** [comment]
> Based on this comment I took `Block` to not be a public class:
> 
> https://github.com/asdf-format/asdf/blob/b8e1759aa9f013283942140c94732a170eb9a703/asdf/block.py#L799-L801
> 
> I do not think this should be public and would rather not include a changelog entry which might cause users to think it's public.
> 

According to [`pep8`](https://peps.python.org/pep-0008/#public-and-internal-interfaces) this is a public interface to the `asdf.block` module, which itself is public. Note that unless specifically specified `__all__` is assumed to be all module level objects without a prefixed `_`. Therefore I consider `Block` to be public.

The docstring/comment you listed is not easily accessible without actually diving into the source code, so I think its unreasonable to expect users to know that it should remain private.

> Are there changes that should be made to make this more explicit?

The obvious solution is to make either the `block` module private or the `Block` class private.

File: asdf/block.py [hunk 1]
```diff
@@ -1005,6 +1009,9 @@ def read(self, fd, past_magic=False, validate_checksum=False):
             If `True`, validate the data against the checksum, and
             raise a `ValueError` if the data doesn't match.
         """
+        if self._data_callback is not None:
```
Comment (perrygreenfield) [inline asdf/block.py:1012]: Shouldn't the docstring for this method also point this out?

**Review by perrygreenfield — COMMENTED** [review]
Only needs an update to a docstring in my opinion.

**Review by perrygreenfield — APPROVED** [review]
LGTM

File: asdf/block.py (hunk 1 above)
Comment (braingram) [reply asdf/block.py:1029]: Thanks! I added some descriptions to the docstring. Let me know if this looks good to you.

**Pull request merged**